### ISSUE TITLE
Block Parameter Scoping (vs Ruby)

### ISSUE THREAD

**Issue by bppr** [jruby/jruby#2960]
I found that if I use parens around a set of block parameters, JRuby seems to be re-assigning the existing variable in the containing scope. MRI keeps the block parameter local to the block. This only has been an issue in practice if there is a defined variable with the block param's name, but thought it was worth bringing up.

Here's some minimal example code to show the differences in behavior between JRuby and MRI.

```ruby
def foo(key)  # no parens around block params
  { a: "b", c: "d" }.map { |key, value| [key, value] }
  return key
end

def bar(key)  # note the presence of parens around block params
  { a: "b", c: "d" }.map { |(key, value)| [key, value] }
  return key
end

# Ruby (tested with 1.9.3-p484 and 2.2.0)
puts foo(:hello)  # => :hello
puts bar(:hello)  # => :hello

# JRuby (1.7.20)
puts foo(:hello)  # => :hello
puts bar(:hello)  # => :c
```

**Comment by enebo** [comment]
Just verified this does not affect jruby 9k which has same behavior as MRI

**Comment by headius** [comment]
I've committed a fix for 1.7 that we may want to mimic in 9k.

Currently 9k works around bad parser output in IRBuilder by forcing all variables for a multiple assignment to be block-local if they're arguments to the block. The changes I made to 1.7 fixes (or at least improves) the parser output to force the proper scoping of these variables at parse time. I have not ported it forward because of the other workarounds done in IR that would need to be removed or modified.
